org.eclipse.jdt.core.javac/src/org/eclipse/jdt/internal/codeassist/RelevanceUtils.java

```diff
@@ -131,6 +131,10 @@ static int computeRelevanceForExpectingType(ITypeBinding proposalType, ExpectedT
 						relevance = CompletionEngine.R_EXPECTED_TYPE;
 					}
 				}
+				if ("Ljava/lang/Object;".equals(expectedType.getKey()) && proposalType.isPrimitive()) {
+					// can be autoboxed
+					relevance = CompletionEngine.R_EXPECTED_TYPE;
+				}
```

Comment:
Shouldn't this be covered at binding level instead? Basically one of the former lines about subtype/supertype could cover it, assuming binding.isSubtypeCompatible() is implemented better.
But then the question becomes how is `isSubtypeCompatible` supposed to behave for primitive? Are there binding tests covering it?

Reply:
The Javadoc claims it's supposed to follow the rules written in out in the Java 3 Java Language Specification. I can't find that document easily. Java 3 was before autoboxing existed (it was introduced in Java 5). In the Java 24 specification, they don't specify the any relation between primitive types and other types.

So yeah, I'll have to look for some tests I guess.

Reply:
OK, I see similar code is actually implemented in the legacy completion engine with something like `convertedType.isCompatibleWith(targetType, this)`. I don't know if such a condition would be better in practice (I suspect not so much, in which case no need to bother further with it)

Reply:
I made the change in `JavacTypeBinding` and it didn't cause any tests to fail, so I guess we could do it either way. I think I'll stick with this way unless there is a compelling reason to move the auto boxing logic into the type binding, since I think the current subtype compatible implementation lines up with the "subtype compatible according to the JLS" wording better.